### ISSUE TITLE
IDE0078 should be configurable to only apply to type checks

### ISSUE THREAD

**Issue by bzd3y** [dotnet/roslyn#72711]
## Summary

Currently IDE0078 seems pretty aggressive in suggestion pattern matching.

## Background and Motivation

IDE0078 suggests replacing comparisons using operators with pattern matching. That seems overly aggressive if it can't be configured. It seems like it only does it when there are multiple terms in the expression, which leads to inconsistency. A simple comparison that uses `==` doesn't get triggered, but something with two `==` and an `||` between does. So sometimes it gets suggested to use pattern matching and sometimes it doesn't for the exact same semantics.

## Proposed Feature

I think letting this rule be configurable would help. An example is:

```ini
csharp_style_prefer_pattern_matching = only_for_type_checks
```

And with that turned on, it only suggests pattern matching for type checks.

In the code I am working on, IDE0078 is actually triggering on something that is already using pattern matching for type checks, but it is using repeated terms with `is` with `||` in between instead of just `or`. So I like that it suggests to simplify that expression into one `is` `is type1 or type2 or type3 ...`. But to get that, I also have to have it suggest basically every single use of `==` in a compound boolean expression also be replaced with pattern matching.

## Alternative Designs

I guess the following would be a compiler feature, so I get it is probably a long shot, but if others thing it is a good idea maybe I should make a new issue for it.

But it would be nice to be able to use `==` in pattern matching like the other operators as that would also help with consistency and in my opinion readability/clarity. In my mind "`is`" makes sense for checking a type and `==` makes sense for comparing values. With that being said, I realize that `is` has been used for general pattern matching and that isn't going to get changed (not that I would really want to anyway). But an improvement to that would be more consistent syntax where the `is` basically starts a pattern matching expression and all of the comparison operators (or anything with a `bool` type?) can be used after that like normal. For example, consider:

```csharp
if (something == 0 || something >= 10) ...
```

That gets a suggestion like:

```csharp
if (something is 0 or >= 10) ...
```

Would be clearer/more readable (in my opinion, anyway) as something like:

```csharp
if (something is == 0 or >= 10) ...
```

And of course the old syntax could still be valid. After all `is` and `==` are already essentially redundant here anyway. But this way you wouldn't have `is`s and `is not`s (which is longer than `!=`...) mixed with your pattern matching expressions. If this were a possibility then I wouldn't mind the suggestion to replace most comparisons with pattern matching as much.

**Comment by dibarbet** [comment]
(Triaging) tagging @CyrusNajmabadi to see if this is something we want to support in the analyzer

**Comment by CyrusNajmabadi** [comment]
> That gets a suggestion like:

The suggestion looks appropriate to me.  I think teh analyzer is doing an appropriate job here, and limiting just to type checks isn't needed.  

>  In my mind "is" makes sense for checking a type and == makes sense for comparing values

I don't think we want the analyzers to try to match general intuitions like this.  From the lang perspective `is` is highly appropriate for checks, period.  It's not special for types, nor would we want to limit it just to that.

**Comment by dibarbet** [comment]
Going to close this as by design then since we're happy with the behavior.

**Comment by bzd3y** [comment]
@CyrusNajmabadi I'm not saying to limit it. I'm saying it could be configurable. There are other rules that are configurable to limit how aggressively they produce suggestions or to just focus on different things.

So are the comparison operators obsolete now? Why or when would they be used if the analyzers are going to suggest using pattern matching?

**Comment by CyrusNajmabadi** [comment]
> So are the comparison operators obsolete now?

No :-)

> Why or when would they be used if the analyzers are going to suggest using pattern matching?

For all the cases where patterns don't work.  For example, a pattern can only check against constant values.  So if you want to compare against a normal non -constant, you couldn't use a pattern.
